src/smc-webapp/codemirror/custom-modes.coffee

```diff
@@ -89,6 +90,35 @@ CodeMirror.defineMode "sagews", (config) ->
 
     return CodeMirror.smc_multiplexing_mode(CodeMirror.getMode(config, "python"), options...)
 
+CodeMirror.defineMode "rmd", (config) ->
+    # derived from the sagews modes with some additions
+    modes = _.clone(_.object(sagews_decorator_modes))
+    modes['fortran95'] = modes['fortran']
+    modes['octave'] = 'octave'
+    modes['bash'] = modes['sh']
+
+    options = []
+
+    # blocks (ATTN ruby before r!)
+    # all engine modes: names(knitr::knit_engines$get())
+    for name in  ['ruby', 'r', 'python', 'octave', 'fortran95', 'fortran',  'octave', 'bash', 'go', 'julia', 'perl']
+        mode = modes[name]
+        open = new RegExp("```\\s*{#{name}[^}]*?}")
```

Comment:
the block to highlight is a markdown triple backticks codeblock, which has `{r [, options ...]}` next to it in the first line. it could also be indented when used in a list, which this also matches, but that's sort of not a really good idea to do in general.